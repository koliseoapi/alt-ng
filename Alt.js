import ActionDispatcher from "./src/ActionDispatcher";
import createActionWrapper from "./src/createActionWrapper";
import { isString } from "./src/utils";

// pass-through dummy action handler
function forwardValue(...args) {
  return args.length > 1 ? args : args[0];
}

/**
 * process the "generate" field of an Actions object. Adds a new pass-through
 * method with each name from the generated values
 */
function generateActions({ generate, ...actions }) {
  return !generate
    ? actions
    : Object.assign(
        generate.reduce((obj, name) => {
          obj[name] = forwardValue;
          return obj;
        }, {}),
        actions
      );
}

export default class Alt {
  constructor(config) {
    this.actions = {};
    this.stores = {};
    this.dispatcher = new ActionDispatcher();

    // our internal reference to stores
    this._stores = {};
  }

  /**
   * Creates the actions for the passed Actions object. If it includes a generated attribute,
   * it will generate pass-through methods with all these names
   */
  createActions(namespace, actions) {
    if (process.env.NODE_ENV !== "production") {
      if (!namespace || !isString(namespace)) {
        throw new Error("Missing namespace");
      }
      if (!actions || !Object.keys(actions).length) {
        throw new Error("actions argument is missing");
      }
      if (this.actions[namespace]) {
        throw new Error(
          `An Actions instance already exists with name ${namespace}`
        );
      }
    }

    actions = generateActions(actions);
    const actionWrappers = Object.keys(actions).reduce((obj, actionName) => {
      const type = `${namespace}/${actionName}`;
      const actionWrapper = createActionWrapper(
        this.dispatcher,
        type,
        actions[actionName]
      );
      obj[actionName] = actionWrapper;
      return obj;
    }, {});

    return (this.actions[namespace] = actionWrappers);
  }

  createStore(displayName, store) {
    if (process.env.NODE_ENV !== "production") {
      if (!displayName || !isString(displayName)) {
        throw new Error("Missing displayName");
      }
      if (!store || !store.eventSource) {
        throw new Error("Missing Store, or not an instance of Store");
      }
      if (this.stores[displayName]) {
        throw new Error(`Store already defined with name ${displayName}`);
      }
    }

    store.displayName = displayName;
    store.dispatcher = store.dispatcher
      ? this.dispatcher.merge(store.dispatcher)
      : this.dispatcher;

    this._stores[displayName] = store;
    // the public interface, just getState and subscribe
    return (this.stores[displayName] = {
      displayName,

      getState() {
        return store.state;
      },

      subscribe(callback) {
        return store.subscribe(callback);
      },

      remove: () => {
        this._stores = this._stores.filter(
          store => store.displayName !== displayName
        );
        delete this.stores[displayName];
      },

      clearListeners() {
        store.clearListeners();
      }
    });
  }
}
