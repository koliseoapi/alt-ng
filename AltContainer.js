/**
 * Will pass the state of a store as properties to the nested components.
 * If there is more than one store, the stores themselves will be passed
 */
import React from "react";
import PropTypes from "prop-types";

export default class AltContainer extends React.Component {
  constructor(props) {
    super(props);
    const { store, stores } = props;

    if (process.env.NODE_ENV !== "production") {
      if (!stores && !store) {
        throw new Error("Must define either store or stores");
      }
      if (stores && store) {
        throw new Error("Cannot define both store and stores");
      }
    }
    this.onStoreChange = this.onStoreChange.bind(this);
    this.state = this._reduceState();
    this.storeListeners = [];
  }

  /*
  TODO: maybe use componentDidUpdate instead (is this really necessary?)
  componentWillReceiveProps(nextProps) {
    // todo: compare with the current properties, maybe we don't need to do anything
    this._destroySubscriptions();
    this.setState(this._reduceState(nextProps));
    this._registerStores(nextProps);
  }
  */

  componentDidMount() {
    this._registerStores(this.props);
  }

  componentWillUnmount() {
    this._destroySubscriptions();
  }

  _reduceState() {
    const { store, stores, mergeFunc } = this.props;
    return store ? store.getState() : mergeFunc(stores);
  }

  _registerStores({ store, stores }) {
    stores = stores || [store];
    stores.forEach(this._addSubscription.bind(this));
  }

  _destroySubscriptions() {
    this.storeListeners.forEach(storeListener => storeListener.dispose());
    this.storeListeners = [];
  }

  _addSubscription(store) {
    this.storeListeners.push(store.subscribe(this.onStoreChange));
  }

  onStoreChange() {
    this.setState(this._reduceState(this.props));
  }

  render() {
    const {
      element,
      children,
      store,
      stores,
      mergeFunc,
      ...nodeProps
    } = this.props;
    return React.createElement(
      element,
      nodeProps,
      React.Children.map(children, child => {
        const childrenProps = Object.assign({}, this.state, child.props);
        return React.cloneElement(child, childrenProps);
      })
    );
  }
}

const storeShape = PropTypes.shape({
  getState: PropTypes.func.isRequired
});
AltContainer.propTypes = {
  // specify either store (a single store) or stores (array of stores)
  store: storeShape,
  stores: PropTypes.arrayOf(storeShape),

  // when specifying multiple stores, this is the function that will be called to merge the states of all stores.
  // By default, Object.assign() will be invoked with all the states.
  mergeFunc: PropTypes.func,

  // when rendering multiple children, a parent element will be created. Default is "div"
  element: PropTypes.string
};

AltContainer.defaultProps = {
  element: "div",
  mergeFunc: function(stores) {
    return stores.reduce(function(state, store) {
      return Object.assign(state, store.getState());
    }, {});
  }
};
