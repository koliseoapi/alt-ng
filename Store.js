import EventSource from "./src/EventSource";
import { isMutableObject } from "./src/utils";
import ActionDispatcher from "./src/ActionDispatcher";

class Store {
  constructor() {
    this.eventSource = new EventSource();
    this.state = {};

    // to be set by Alt.createStore()
    // this.dispatcher
  }

  bindActions(actions) {
    Object.keys(actions).forEach(name => {
      const type = actions[name].type;
      const handler = this[name] || this.otherwise;
      handler && this.bindAction(type, handler);
    });
  }

  bindAction(action, handlerMethod) {
    // bind the method to the instance
    const handler = handlerMethod.bind(this);
    const type = action.type ? action.type : action;
    handler.store = this;
    if (!this.dispatcher) {
      // create a temporary action dispatcher, will be merged into a single instance later
      this.dispatcher = new ActionDispatcher();
    }
    this.dispatcher.subscribe(type, handler);
  }

  // where keys are methods in the Store, and the values are the actions
  bindListeners(listeners) {
    Object.keys(listeners).forEach(name => {
      const handler = this[name];
      if (!handler) {
        throw new Error(`${name} is not defined in Store`);
      }

      let values = listeners[name];
      values = Array.isArray(values) ? values : [values];
      values.forEach(action => this.bindAction(action, handler));
    });
  }

  subscribe(callback) {
    return this.eventSource.subscribe(callback);
  }

  setState(nextState) {
    if (isMutableObject(this.state)) {
      this.state = Object.assign({}, this.state, nextState);
    } else {
      this.state = nextState;
    }
  }

  preventDefault() {
    this._triggerChange = false;
  }

  // emit a change in the Store state
  // @param value the value to be emitted, default is this.state
  emitChange(value = this.state) {
    this.eventSource.publish(value);
  }

  // remove all listeners
  clearListeners() {
    this.eventSource.clear();
  }
}

export default Store;
