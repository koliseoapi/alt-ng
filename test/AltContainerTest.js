import Alt from "../Alt";
import Store from "../Store";
import React from "react";
import AltContainer from "../AltContainer";
import renderer from "react-test-renderer";

const alt = new Alt();
const actions = alt.createActions("Actions", { generate: ["sup"] });

class TestStore extends Store {
  constructor() {
    super();
    this.bindActions(actions);
    this.state = { x: null };
  }

  sup(x) {
    this.setState({ x });
  }
}

class TestStore2 extends Store {
  constructor() {
    super();
    this.bindActions(actions);
    this.state = { y: null };
  }

  sup(y) {
    this.setState({ y });
  }
}

const testStore = alt.createStore("testStore", new TestStore(actions));
const testStore2 = alt.createStore("testStore2", new TestStore2(actions));

describe("AltContainer", () => {
  let wrapper;

  it("requires either store or stores", () => {
    expect(() =>
      renderer.create(
        <AltContainer>
          <div />
        </AltContainer>
      )
    ).toThrowError("Must define either store or stores");
    expect(() =>
      renderer.create(
        <AltContainer store={testStore} stores={[testStore]}>
          <div />
        </AltContainer>
      )
    ).toThrowError("Cannot define both store and stores");
  });

  it("renders HTML", () => {
    expect(
      renderer
        .create(
          <AltContainer store={testStore}>
            <span className="foo" />
          </AltContainer>
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it("renders multiple children with keys", () => {
    expect(
      renderer
        .create(
          <AltContainer store={testStore}>
            <div key="1" />
            <div key="2" />
            <div key="3" />
          </AltContainer>
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it("store state is propagated as props", () => {
    let tree = renderer.create(
      <AltContainer store={testStore}>
        <div key="div1" className="div1" />
        <div key="div2" className="div2" />
      </AltContainer>
    );
    actions.sup("hello");
    expect(tree.toJSON()).toMatchSnapshot();
    actions.sup("bye");
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("multiple store states are merged", () => {
    let tree = renderer.create(
      <AltContainer stores={[testStore, testStore2]}>
        <div className="xxx" />
      </AltContainer>
    );
    actions.sup("foobar");
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("should apply changes to properties", function() {
    const node = document.createElement("div");
    let tree = renderer.create(
      <AltContainer store={testStore}>
        <div className="xxx" />
      </AltContainer>,
      node
    );
    actions.sup("strike1");
    expect(tree.toJSON()).toMatchSnapshot();
    renderer.create(
      <AltContainer store={testStore2}>
        <div className="xxx" />
      </AltContainer>,
      node
    );
    actions.sup("strike2");
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("keeps refs in children", function() {
    const onRef = jest.fn(() => document.createElement("div"));
    class Container extends React.Component {
      render() {
        return (
          <AltContainer store={testStore}>
            <span ref={onRef} />
          </AltContainer>
        );
      }
    }
    const tree = renderer.create(<Container />);
    actions.sup("hello");
    expect(tree.toJSON()).toMatchSnapshot();
    expect(onRef).toBeCalledTimes(1);
  });
});
