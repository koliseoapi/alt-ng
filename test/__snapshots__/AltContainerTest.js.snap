// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AltContainer keeps refs in children 1`] = `
<div>
  <span
    x="hello"
  />
</div>
`;

exports[`AltContainer multiple store states are merged 1`] = `
<div>
  <div
    className="xxx"
    x="foobar"
    y="foobar"
  />
</div>
`;

exports[`AltContainer renders HTML 1`] = `
<div>
  <span
    className="foo"
    x={null}
  />
</div>
`;

exports[`AltContainer renders multiple children with keys 1`] = `
<div>
  <div
    x={null}
  />
  <div
    x={null}
  />
  <div
    x={null}
  />
</div>
`;

exports[`AltContainer should apply changes to properties 1`] = `
<div>
  <div
    className="xxx"
    x="strike1"
  />
</div>
`;

exports[`AltContainer should apply changes to properties 2`] = `
<div>
  <div
    className="xxx"
    x="strike2"
  />
</div>
`;

exports[`AltContainer store state is propagated as props 1`] = `
<div>
  <div
    className="div1"
    x="hello"
  />
  <div
    className="div2"
    x="hello"
  />
</div>
`;

exports[`AltContainer store state is propagated as props 2`] = `
<div>
  <div
    className="div1"
    x="bye"
  />
  <div
    className="div2"
    x="bye"
  />
</div>
`;
